pkg/admission/validatingwebhook/plugin_test.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validatingwebhook
+
+import (
+	"testing"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const clusterName = logicalcluster.Name("test-cluster")
+
+func TestSetClusterAnnotation(t *testing.T) {
+	tests := []struct {
+		name string
+		in   *corev1.ConfigMap
+	}{
+		{
+			name: "no annotations",
+			in: &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{},
+			},
+		},
+		{
+			name: "with annotations",
+			in: &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						"foo": "bar",
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			origAnnotations := test.in.GetAnnotations()
+
+			undo := SetClusterAnnotation(test.in, clusterName)
+			assert.NotNil(t, undo)
+
+			require.NotNil(t, test.in.GetAnnotations())
```

Comment:
Zero-value of `GetAnnotations()` not being `nil` even though you don't set it to anything in an empty `ObjectMeta` is enforced by what? Doesn't Go semantic not do that? https://go.dev/play/p/vYJPy8IEjXq

Reply:
`SetClusterAnnotation` is supposed to set annotations if no annotations are set initially:
https://github.com/kcp-dev/kcp/pull/3229/files#diff-21219bad7e49f362d06f1668718121ef01005c2a0a8133188b31a32b9944b097R215

Reply:
Ah, I see - d'oh. The only difference I can see in your impl which might be unintended is that if originally annotations were `nil`, does your `undo` return it to `nil`? Does it matter?

Reply:
I was thinking about resetting to nil if the map would be left empty as well, but I think it doesn't matter.
On the wire it would be omitted either way looking into the json encoding:

https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/encoding/json/encode.go;l=73-76
https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/encoding/json/encode.go;l=704
https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/encoding/json/encode.go;l=307-310

Reply:
That was my suspicion, thank you.